tools/code-analyzer/src/git.ts

```diff
@@ -5,12 +5,135 @@ import { CliUx } from '@oclif/core';
 import { execSync } from 'child_process';
 import { join } from 'path';
 import { tmpdir } from 'os';
-import { readFileSync } from 'fs';
+import { readFileSync, rmSync } from 'fs';
+import { simpleGit } from 'simple-git';
+import { v4 } from 'uuid';
 
 /**
  * Internal dependencies
  */
 import { startWPEnv, stopWPEnv, isValidCommitHash } from './utils';
+/**
+ * Clone a git repository.
+ *
+ * @param {string} repoPath- the path (either URL or file path) to the repo to clone.
+ * @param          repoPath
+ * @return {string} the path to the cloned repo.
+ */
+export const cloneRepo = async ( repoPath: string ) => {
+	const tempFolderName = v4();
+	const git = simpleGit( { baseDir: tmpdir() } );
+	await git.clone( repoPath, tempFolderName );
+
+	return tempFolderName;
+};
+
+/**
+ * Do a git diff of 2 commit hashes (or branches)
+ *
+ * @param {string} baseDir - baseDir that the repo is in
+ * @param {string} hashA   - either a git commit hash or a git branch
+ * @param {string} hashB   - either a git commit hash or a git branch
+ * @return {string} - diff of the changfiles between the 2 hashes
+ */
+export const diffHashes = ( baseDir: string, hashA: string, hashB: string ) => {
+	const git = simpleGit( { baseDir } );
+	return git.diff( [ `${ hashA }..${ hashB }` ] );
+};
+
+/**
+ * Determines if a string is a commit hash or not.
+ *
+ * @param {string} ref - the ref to check
+ * @return {boolean} whether the ref is a commit hash or not.
+ */
+const refIsHash = ( ref: string ) => {
+	return /^[0-9a-f]{7,40}$/i.test( ref );
+};
+
+/**
+ * Get the commit hash for a ref (either branch or commit hash). If a validly
+ * formed hash is provided it is returned unmodified.
+ *
+ * @param {string} baseDir - the dir of the git repo to get the hash from.
+ * @param {string} ref     -    Either a commit hash or a branch name.
+ * @return {string} - the commit hash of the ref.
+ */
+export const getCommitHash = async ( baseDir: string, ref: string ) => {
+	const isHash = refIsHash( ref );
+
+	// If its not a hash we assume its a branch
+	if ( ! isHash ) {
+		return simpleGit( { baseDir } ).revparse( [ ref ] );
+	}
```

Comment:
Not sure why, but the script has trouble with branches as opposed to hashes.

```
➜  woocommerce git:(dev/use-simple-git) ./tools/code-analyzer/bin/dev analyzer dev/use-simple-git
Error: Unable to create diff. Check that git repo, base hash, and compare hash all exist.
 Error: fatal: Unable to create '/private/var/folders/2p/zg7bwl6d4mzb1w5m42p48yn80000gn/T/cea3a8ee-c47f-4610-ad17-828d33d6c9fd/.git/index.lock': File exists.

Another git process seems to be running in this repository, e.g.
an editor opened by 'git commit'. Please make sure all processes
are terminated then try again. If it still fails, a git process
may have crashed in this repository earlier:
remove the file manually to continue.

    at Object.error (/Users/paulsealock/Workspace/woocommerce/node_modules/.pnpm/@oclif+core@1.3.4/node_modules/@oclif/core/lib/errors/index.js:29:15)
    at error (/Users/paulsealock/Workspace/woocommerce/node_modules/.pnpm/@oclif+core@1.3.4/node_modules/@oclif/core/lib/command.js:75:23)
    at generateDiff (/Users/paulsealock/Workspace/woocommerce/tools/code-analyzer/src/git.ts:125:4)
    at async Analyzer.run (/Users/paulsealock/Workspace/woocommerce/tools/code-analyzer/src/commands/analyzer/index.ts:90:16)
    at async Analyzer._run (/Users/paulsealock/Workspace/woocommerce/node_modules/.pnpm/@oclif+core@1.3.4/node_modules/@oclif/core/lib/command.js:52:22)
    at async Config.runCommand (/Users/paulsealock/Workspace/woocommerce/node_modules/.pnpm/@oclif+core@1.3.4/node_modules/@oclif/core/lib/config/config.js:226:25)
    at async Object.run (/Users/paulsealock/Workspace/woocommerce/node_modules/.pnpm/@oclif+core@1.3.4/node_modules/@oclif/core/lib/main.js:73:5)
```

Reply:
Thanks I'll dig into why this is happening

Reply:
Hmm, in testing I found that the db schema command sometimes blitzes the repo altogether. I wonder if subsequent calls are causing this issue because by default this clones from the existing repo. 🤔

Reply:
There were a few bugs I hadn't caught on to here.

1. simple-git can't run parallel `git checkout` commands, because of how concurrency works. `Promise.all` wasn't going to work there so I fixed that.

2. I noticed that when you do a clone from a local repo the remote is set to be the file path of the original. Which makes sense but I didn't think of that when cloning. So now if you clone from local, I copy the git config over too in that case, so that the remote is set the same. (I learned a lot about git today!)

Now when testing the command you've suggested here I see that it works well for me. See https://github.com/woocommerce/woocommerce/pull/33837/commits/fe05294d9d3e3caa2e51d876731f1a5debd961b0